### ISSUE TITLE
SuperTuxKart 0.8.1

### ISSUE THREAD

**Issue by kas1e** [ptitSeb/gl4es#130]
So, after i made proper irrlicht port, and it seems all works fine now , and after VBO things was added, i back to SuperTuxKart 0.8.1 in hope to make it works on aos4 :)

Currently have some general questions: 

1). Is on running, you have in log from irr_driver: "Too old GPU; using fixed pipeline" ?

2). Is for you main loading screen take a bit of time ? I.e. about minute or two ?

**Comment by ptitSeb** [comment]
For the gpu message, I think you need to use LIBGL_GL=21

Loading screen is a bit long, but I don't think it take more than a minute.

**Comment by kas1e** [comment]
seems LIBGL_GL 21 didn't make it .. but need recheck, maybe i do something wrong ..

**Comment by kas1e** [comment]
Rechecked, LIBGL_GL 21 for sure didn't help, still have those words about "using fixed pipeline".

**Comment by ptitSeb** [comment]
I checked of the Pandora, and in fact, it needs `LIBGL_NPOT=2`. Also, on the pandora, I also need `LIBGL_FORCENPOT=1` to avoid missing texture.
But maybe on Amiga this is not the case? Could this "GL_ARB_texture_non_power_of_two" or this "GL_OES_texture_npot" extension be supported on the Amiga but not exposed?

**Comment by kas1e** [comment]
What you mean by "missing textures" ? I seems to have all texture in place (but still fighting with unix pathes issues), but at least seems i can see verything. Just at moment found a  crash when i in the "choice a cart" menu, then on some i crashes in some texture related function, maybe that what you mean ?

**Comment by ptitSeb** [comment]
No, I don't mean a crash, I mean a missing texture: a texture with NPOT dimension (Non Power Of Two) with Mipmap or with GL_REPEAT, if the hardware don't support it, means the texture is incomplete, and will not show correctly. But if you have everything, that may mean NPOT texture are fully supported (that would make sense since the a Desktop class graphic card).

**Comment by kas1e** [comment]
But for "Too old GPU; using fixed pipeline" are you have it too ? Or for you it skips once you set LIBGL_GL=21 ? Maybe its something about default configs ..

**Comment by ptitSeb** [comment]
No, I don't need `LIBGL_GL=21` that was a mistake. Just `LIBGL_NPOT=2` should be enough to have the message "irr_driver: GLGL supported."

**Comment by kas1e** [comment]
I somehow was able to start some track , and even with Fixed Pipeline it was more or less fast. I mean, really playable, better than old supertuxkart for sure. 

As for LIBGL_NPOT=2 , yeah, that did the trick, through as expected "GLSL shader failed to compile" , but with that one i can deal later , firstly those damn unix pathes.

**Comment by kas1e** [comment]
And should to say crash is pretty strange, before it crashes, it bring me to console:

Irrlicht: FBO has one or severla incomplete image attachements
Irrlicht: FBO error

And stack trace looks like this (Through without normal debug symbols):

```
Symbol info:
Instruction pointer 0x7F2AE578 belongs to module "supertuxkart" (PowerPC) 
Symbol: _ZN3irr5video13COpenGLDriver11draw2DImageEPKNS0_8ITextureERKNS_4core4rectIiEES9_PS8_PKNS0_6SColorEb + 0x2C in section 1 offset 0x003CF574

Stack trace:
    supertuxkart:_ZN3irr5video13COpenGLDriver11draw2DImageEPKNS0_8ITextureERKNS_4core4rectIiEES9_PS8_PKNS0_6SColorEb()+0x2c (section 1 @ 0x3CF574)
    supertuxkart:_ZN9GUIEngine4Skin14drawIconButtonERKN3irr4core4rectIiEEPNS_6WidgetEbb.constprop.255()+0x1bc (section 1 @ 0xCEEB0)
    supertuxkart:_ZN9GUIEngine4Skin13process3DPaneEPN3irr3gui11IGUIElementERKNS1_4core4rectIiEEb()+0x4ac (section 1 @ 0xD0B74)
    supertuxkart:_ZN9GUIEngine4Skin24draw3DButtonPaneStandardEPN3irr3gui11IGUIElementERKNS1_4core4rectIiEEPS8_()+0x84 (section 1 @ 0xD183C)
    supertuxkart:_ZN3irr3gui10CGUIButton4drawEv()+0x68c (section 1 @ 0x60B930)
    supertuxkart:_ZN3irr3gui15CGUIEnvironment7drawAllEv()+0x1a4 (section 1 @ 0x48D72C)
    supertuxkart:_ZN9GUIEngine6renderEf()+0x43c (section 1 @ 0xA937C)
    supertuxkart:_ZN9IrrDriver6updateEf()+0x858 (section 1 @ 0x6F9DC)
    supertuxkart:_ZN8MainLoop3runEv()+0x2b4 (section 1 @ 0x196208)
    supertuxkart:main()+0x8bc (section 1 @ 0x194D84)
```

Never seen anything of that sort ? Its just when i start to move mouse cursor in the "kart choice" menu, over all of them. Few of them works, other ones crashes like that.

**Comment by kas1e** [comment]
Also i measure loading time till first menu showups : 1 min 40 secs. Is it faster for you on pandora ?

**Comment by kas1e** [comment]
Oh, and about that crash, one time i catch not only those FBO errors, but it also bring me that:

```
Irrlicht: FBO has one or severla incomplete image attachements
Irrlicht: FBO error
main: Exception cught : std::bad_alloc.
main: Aborting SuperTuxKart
```

**Comment by ptitSeb** [comment]
The most demanding track, at the beggining at least, is the "Scotland" one (maybe Subsea is quite heavy too, I don't remember). The less demanding is the "Shifting Sand" one. VBO helps here, even on the Pandora, so I guess the effect is very sensitive on the Amiga.

And the shaders, yeah, it's not a big deal if they don't work, it mostly a Blur effect when using turbo that is not that usefull.

Good luck with the unix2amiga path conversions!

**Comment by ptitSeb** [comment]
Yeah, it's faster than that on the Pandora. I'll make some timing later (too late now).

About the FBO, strange, it works on the Pandora. You can maybe activate debug log in `src/framebuffer.c` to see what is happening in FBO creation & use.

**Comment by kas1e** [comment]
I deal with unix patches (mostly), just one little thing left, but it hard for me to debug, because 50% of time i crashes on the stage where need to choice a car :( If i didn't crash there, then i can choice tracks, play in them, etc ,etc. So  will try to activate debug in src/framebuffer.c now

**Comment by kas1e** [comment]
See the speed i have on amigaos (when i didn't crash on kar-choice-screen):

https://youtu.be/h37La7ylaZo

**Comment by kas1e** [comment]
That 1024*768, by default it 800*600 and about twice faster

**Comment by ptitSeb** [comment]
The speed seems correct. Well playable. And I guess you can still remove some effects in the options.

One thing I noticed is that the FBO seems to have the wrong RGBA channels: the rotating car is Blue where it should be Red. So maybe the texture created for the FBO is in a GL_BGRA format and the OGLES2/Warp3D driver doesn't like that?

You need that trace (IIRC, there are some details about attached textures and renderbuffer with the debug traces of framebuffer.c)

**Comment by kas1e** [comment]
Enabled debug in ```src/framebuffer.c```, that what i have:

```
glGenerateMipmap(GL_TEXTURE_2D)
glGenerateMipmap(GL_TEXTURE_2D)
glGenFramebuffers(1, 0x58e7e738)
glBindFramebuffer(GL_FRAMEBUFFER, 256), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, 568, 0) glstate->fbo.current_fb=256 (draw=256, read=256)
found texture, glname=568, size=512x512(512x512), format/type=GL_BGRA/GL_UNSIGNED_BYTE
Attach Texture 568 to FBO 256 as Attachement GL_COLOR_ATTACHMENT0
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=256 (draw=256, read=256)
glGenRenderbuffers(1, 0x58e4fb04)
glBindRenderbuffer(GL_RENDERBUFFER, 256), binded Fbo=0
glRenderbufferStorage(GL_RENDERBUFFER, 0x81A5, 512, 512)
glBindFramebuffer(GL_FRAMEBUFFER, 256), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, 256)
glCheckFramebufferStatus(0x8D40)=0x8CD5
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=256 (draw=256, read=256)
glBindFramebuffer(GL_FRAMEBUFFER, 256), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=256 (draw=256, read=256)
glDeleteRenderbuffer(1, 0x58e4fb04)
glDeleteFramebuffers(1, 0x58e7e738), framebuffers[0]=256
glGenFramebuffers(1, 0x58e7e738)
glBindFramebuffer(GL_FRAMEBUFFER, 257), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, 569, 0) glstate->fbo.current_fb=257 (draw=257, read=257)
found texture, glname=569, size=512x512(512x512), format/type=GL_BGRA/GL_UNSIGNED_BYTE
Attach Texture 569 to FBO 257 as Attachement GL_COLOR_ATTACHMENT0
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=257 (draw=257, read=257)
glGenRenderbuffers(1, 0x58e4f4cc)
glBindRenderbuffer(GL_RENDERBUFFER, 257), binded Fbo=0
glRenderbufferStorage(GL_RENDERBUFFER, 0x81A5, 512, 512)
glBindFramebuffer(GL_FRAMEBUFFER, 257), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, 257)
glCheckFramebufferStatus(0x8D40)=0x8CD5
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=257 (draw=257, read=257)
glBindFramebuffer(GL_FRAMEBUFFER, 257), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=257 (draw=257, read=257)
glDeleteRenderbuffer(1, 0x58e4f4cc)
glDeleteFramebuffers(1, 0x58e7e738), framebuffers[0]=257
glGenFramebuffers(1, 0x58e7e738)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, 570, 0) glstate->fbo.current_fb=258 (draw=258, read=258)
found texture, glname=570, size=512x512(512x512), format/type=GL_BGRA/GL_UNSIGNED_BYTE
Attach Texture 570 to FBO 258 as Attachement GL_COLOR_ATTACHMENT0
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glGenRenderbuffers(1, 0x58e4f4cc)
glBindRenderbuffer(GL_RENDERBUFFER, 258), binded Fbo=0
glRenderbufferStorage(GL_RENDERBUFFER, 0x81A5, 512, 512)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, 258)
glCheckFramebufferStatus(0x8D40)=0x8CD6
[error  ] Irrlicht: FBO has one or several incomplete image attachments
[error  ] Irrlicht: FBO error
glDeleteRenderbuffer(1, 0x58e4f4cc)
glDeleteFramebuffers(1, 0x58e7e738), framebuffers[0]=258 
```

and then crash. Sometime it show that "blue" car, sometime not. Sometime i even can choice another car. But seems some of them always fail like that

**Comment by kas1e** [comment]
Will try to catch also output when it didn't crash

**Comment by kas1e** [comment]
And that is the output when it didn't crash:

```
glGenerateMipmap(GL_TEXTURE_2D)
glGenerateMipmap(GL_TEXTURE_2D)
glGenFramebuffers(1, 0x592932a0)
glBindFramebuffer(GL_FRAMEBUFFER, 256), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, 568, 0) glstate->fbo.current_fb=256 (draw=256, read=256)
found texture, glname=568, size=512x512(512x512), format/type=GL_BGRA/GL_UNSIGNED_BYTE
Attach Texture 568 to FBO 256 as Attachement GL_COLOR_ATTACHMENT0
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=256 (draw=256, read=256)
glGenRenderbuffers(1, 0x592934a4)
glBindRenderbuffer(GL_RENDERBUFFER, 256), binded Fbo=0
glRenderbufferStorage(GL_RENDERBUFFER, 0x81A5, 512, 512)
glBindFramebuffer(GL_FRAMEBUFFER, 256), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, 256)
glCheckFramebufferStatus(0x8D40)=0x8CD5
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=256 (draw=256, read=256)
glBindFramebuffer(GL_FRAMEBUFFER, 256), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=256 (draw=256, read=256)
glDeleteRenderbuffer(1, 0x592934a4)
glDeleteFramebuffers(1, 0x592932a0), framebuffers[0]=256
glGenFramebuffers(1, 0x59292e60)
glBindFramebuffer(GL_FRAMEBUFFER, 257), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, 569, 0) glstate->fbo.current_fb=257 (draw=257, read=257)
found texture, glname=569, size=512x512(512x512), format/type=GL_BGRA/GL_UNSIGNED_BYTE
Attach Texture 569 to FBO 257 as Attachement GL_COLOR_ATTACHMENT0
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=257 (draw=257, read=257)
glGenRenderbuffers(1, 0x594284f4)
glBindRenderbuffer(GL_RENDERBUFFER, 257), binded Fbo=0
glRenderbufferStorage(GL_RENDERBUFFER, 0x81A5, 512, 512)
glBindFramebuffer(GL_FRAMEBUFFER, 257), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, 257)
glCheckFramebufferStatus(0x8D40)=0x8CD5
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=257 (draw=257, read=257)
glBindFramebuffer(GL_FRAMEBUFFER, 257), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=257 (draw=257, read=257)
glDeleteRenderbuffer(1, 0x594284f4)
glDeleteFramebuffers(1, 0x59292e60), framebuffers[0]=257
glGenFramebuffers(1, 0x59292e60)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, 570, 0) glstate->fbo.current_fb=258 (draw=258, read=258)
found texture, glname=570, size=512x512(512x512), format/type=GL_BGRA/GL_UNSIGNED_BYTE
Attach Texture 570 to FBO 258 as Attachement GL_COLOR_ATTACHMENT0
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glGenRenderbuffers(1, 0x5977652c)
glBindRenderbuffer(GL_RENDERBUFFER, 258), binded Fbo=0
glRenderbufferStorage(GL_RENDERBUFFER, 0x81A5, 512, 512)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, 258)
glCheckFramebufferStatus(0x8D40)=0x8CD5
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0)
glBindFramebuffer(GL_FRAMEBUFFER, 0), list=none, glstate->fbo.current_fb=258 (draw=258, read=258)
glBindFramebuffer(GL_FRAMEBUFFER, 258), list=none, glstate->fbo.current_fb=0 (draw=0, read=0) 
```

and so on repeat the last lines , all works. But car is blue, yep.

**Comment by ptitSeb** [comment]
I think I have enough logs now. I'll look at them tonight, I don't have time now.

**Comment by kas1e** [comment]
Comparing logs in meantime, they looks pretty much kind of the same .. The only differences i see in crashed one, is that there is 3 glGenFramebuffers with the same address, while on non-crashed only 2. But that can be of no issue probabaly ..

But you are right, texture format indeed  GL_BGRA , but this one we probably use all the time in gl4es already with all other games ?

**Comment by kas1e** [comment]
Ha, what i noticed now as well, is that in Irrlicht engine  , demo 13.RenderToTexture, there is Cube rotated. And on win32 it is blue, while on amigaos4 it is red !

**Comment by ptitSeb** [comment]
Yeah, I think you need to ask Daniel (and maybe Hans), if he seees an issue in bindding a GL_BRGA texture to an FBO.

You have to remind what bidding a texture to an FBO mean: that means that all the renedring that will be done don't go to the screen, but to the binded FBO, so the drawing will be done directly in the texture.
So maybe GL_BGRA are fine when you read it, but maybe are wrongly handling for "writting" in an FBO.

For a quick test, you can launch the game (or the Irrlicht sample) with `LIBGL_NOBGRA=1` to disable the use of `GL_BGRA`

**Comment by ptitSeb** [comment]
Also, the log trace from the "crash" version looks clean. There is 3 FBO created, exactly in the same way, but the 3rd one didn't worked, for some reason I can't explain.

**Comment by kas1e** [comment]
```LIBGL_NOBGRA 1``` fix issue with irrlicht example for sure : cube now blue as expected, and not red. Also in supertuxkart car is now also red and not blue, when didnt' crash, but it still crashes :(

Did it mean that crash unrelated to handling of BGRA and something completely different ? Is there anything we can enabled in gl4es to see more info, so i can provide it to Daniel / Hans ?

It seems that thing about that differnt colors can be just some endian issue in gl4es or irrlicht engine , right ?

**Comment by ptitSeb** [comment]
> Did it mean that crash unrelated to handling of BGRA and something completely different ? Is there anything we can enabled in gl4es to see more info, so i can provide it to Daniel / Hans ?

Yes. But there isn't any more usefull log you can get from gl4es. In the current logs, the issue is when `glCheckFramebufferStatus(...)` is not `0x8CD5`... (this one means FBO Complete).

> It seems that thing about that differnt colors can be just some endian issue in gl4es or irrlicht engine , right ?
Nope. There isn't much Irrlicht engine and gl4es do here. Texture with GL_BGRA format are supported on Amiga, there are other games that support it, so it's an issue with this format. Once a Texture is binded to an FBO, gl4es (and Irrlicht) has finished their job, and now eveyrhting that is drawn (so by Warp3D) goes in the texture. So the issue is either with OGLES2 when creating the FBO or with Warp3D when creating the FBO (that's my guess).

**Comment by kas1e** [comment]
Ok, thanks for explain. Trying to collect all the data about that crash, and in case you may have few minutes to check (and maybe having idea about), there is proper stack trace with debug info:

```
Stack trace:
    0x590F9168 symbol not available
    [/amiga/SuperTuxKart-0.8.1/lib/irrlicht/source/Irrlicht/COpenGLDriver.cpp:2184] supertuxkart:_ZN3irr5video13COpenGLDriver11draw2DImageEPKNS0_8ITextureERKNS_4core4rectIiEES9_PS8_PKNS0_6SColorEb()+0x78 (section 1 @ 0x3CFB70)
    [/amiga/SuperTuxKart-0.8.1/src/guiengine/skin.cpp:1849] supertuxkart:_ZN9GUIEngine4Skin13process3DPaneEPN3irr3gui11IGUIElementERKNS1_4core4rectIiEEb()+0x4ac (section 1 @ 0xD0B74)
    [/amiga/SuperTuxKart-0.8.1/src/guiengine/skin.cpp:2015] supertuxkart:_ZN9GUIEngine4Skin24draw3DButtonPaneStandardEPN3irr3gui11IGUIElementERKNS1_4core4rectIiEEPS8_()+0x84 (section 1 @ 0xD183C)
    [/amiga/SuperTuxKart-0.8.1/lib/irrlicht/source/Irrlicht/CGUIButton.cpp:244] supertuxkart:_ZN3irr3gui10CGUIButton4drawEv()+0x68c (section 1 @ 0x60BEE0)
    [/amiga/SuperTuxKart-0.8.1/lib/irrlicht/source/Irrlicht/CGUIEnvironment.cpp:318] supertuxkart:_ZN3irr3gui15CGUIEnvironment7drawAllEv()+0x1a4 (section 1 @ 0x48DCDC)
    [/amiga/SuperTuxKart-0.8.1/src/guiengine/engine.cpp:1164] supertuxkart:_ZN9GUIEngine6renderEf()+0x43c (section 1 @ 0xA937C)
    [/amiga/SuperTuxKart-0.8.1/src/graphics/irr_driver.cpp:1731] supertuxkart:_ZN9IrrDriver6updateEf()+0x858 (section 1 @ 0x6F9DC)
    [/amiga/SuperTuxKart-0.8.1/src/main_loop.cpp:164] supertuxkart:_ZN8MainLoop3runEv()+0x2b4 (section 1 @ 0x1967D0)
    [/amiga/SuperTuxKart-0.8.1/src/main.cpp:1536] supertuxkart:main()+0x8bc (section 1 @ 0x19534C)
    native kernel module newlib.library.kmod+0x0000257c
    native kernel module newlib.library.kmod+0x00003294
    native kernel module newlib.library.kmod+0x000037c8
    supertuxkart:_start()+0x170 (section 1 @ 0x1920)
    native kernel module dos.library.kmod+0x0002a5d8
    native kernel module kernel+0x0006b590
    native kernel module kernel+0x0006b5d8
 
```

And there i trace everything from starting till crash (all possible warp3d and ogles2 functions):

http://kas1e.mikendezign.com/aos4/gl4es/games/supertuxkart_081/stk_crash_trace.txt

I didn't see there anything obvious, but at least i can see that it crashes inside of Irrlicht's fucntion ```COpenGLDriver::draw2DImage``` , and crashed line are ```const core::dimension2d<u32>& ss = texture->getOriginalSize();``` , i.e. when it want ot get some original size. 

Maybe it just miss to find the texture somehow and just didn't check wrong return and fail like that ..

**Comment by ptitSeb** [comment]
About the crash in the FBO creation, maybe you can try with `LIBGL_RECYCLEFBO=1` to avoid multie FBO creation / deletions.

**Comment by kas1e** [comment]
Oh yeah , it definataly fix the crash ! Through, rendering of those parts now broken, it have parts of previous memory or something. See:

http://kas1e.mikendezign.com/aos4/gl4es/BGRA/RECYCLEFBO_1.jpg

 Can you check if it works on Pandora when you use LIBGL_RECYCLEFBO=1 ? 

If it will be the same on Pandora, then one route, if it will be fine on Pandora but again only on our side, then another route ..

But at least it some hint for sure !

**Comment by kas1e** [comment]
Also can you explain plz in more deep, what that LIBGL_RECYCLEFBO=1 doing exactly (in coding terms), so Daniel will know where to look at (he think that its can be some ogles2 bug which cause a crash).

**Comment by ptitSeb** [comment]
Ah yes, it's old code that I don't test often. so it mayb patialy (or totaly) broken.
I'll try to test that on sundaynight, I'll probably wont be able to test before.

**Comment by kas1e** [comment]
Aha ok, thanks. Is there anything to explain about LIBGL_RECYCLEFBO doing exactly so Daniel will know where to look at ?

**Comment by kas1e** [comment]
I mean that he understand that ```LIBGL_RECYCLEFBO=1```  its for "to avoid multie FBO creation / deletions." , but what it do exactly, i.e. is it reduce number to some other number of creation, or it use one all the time, or what it doing exactly (so that will help to him to see why it crashes in ogles2, at least in which route to check).

Thanks !

**Comment by ptitSeb** [comment]
The idea is that if an FBO is delete, then it's not really delete and goes on a "pool" of unused FBO. Then if an FBO is recreated and the pool is ot empty, that "unused" FBO is taken from the pool and use.

I had some issue with a game the used Ogre (OSG Mahjongg) and create / delete many FBO in a single frame (probably to create some texture), and some driver didn't react well so I developped this workaround.

**Comment by kas1e** [comment]
Thanks! Forwarded.

In meantime we found and fix color-swap issue. Once you will have time to check how libgl_recyclefbo reacts in stk on pandora, we will have route to follow. Maybe all will be fine on your side, and it will be again issue on our side :)

**Comment by kas1e** [comment]
Found something interesting as well : if i choice a car with a brown color for example, and then start a level, then while i didn't start move , car is red. But then once i start move, color changed on brown.

That with LIBGL_RECYCLEFBO 1 . But that can be again side-effect of LIBGL_RECYCLEFBO itself (if it broken at all in gl4es and its not on our side).  Haven't had time to check if  LIBGL_RECYCLEFBO 1 works in supertuxkart for you ?

**Comment by ptitSeb** [comment]
Ah yes, sorry, I forgot.

I've just tested, and yeah, it's broken. On the Pandora, I just have ... a small non-moving garbaged rectangle instead of the rotating kart...

**Comment by kas1e** [comment]
Oh yeah, that good to know that its not on our side again :) I also have some weird crashes in some levels (on loading), which can be also side effect of that LIBGL_RECYCLEFBO 1 , but can't be sure of course..

**Comment by ptitSeb** [comment]
Ok, I have just pushed a fix for this. It's working fine on the Panora now. (And yeah, previously, it could have make some crash happening)

**Comment by kas1e** [comment]
Oh damn, now, when it start to looks correctly with LIBGL_RECYCLEFBO 1  , it then crashes the same as without ! :(

With the same stack trace, and with the same output in shell:

[error] Irrlicht: FBO has one or several incomplete image attachements
[error] Irrlich: FBO error

Quite strange ! I mean, without fix, it didn't show things correclty, but didn't crashes. Now it start to show things correctly, but then crashes after i move cursor over few cars .. Hm, at least that point out us on something that there is some general issue ..

**Comment by ptitSeb** [comment]
Mmm, maybe FBO needs some pause to be dealt with?

Just a quick test, in `src/framebuffer.c` line 376 (so inside `gl4es_glFramebufferTexture2D`) add
```
LOAD_GLES(glFlush);
gles_glFlush();
```
To flush all GL operation before attaching a texture to the FBO.

**Comment by kas1e** [comment]
Tried , nope, the same :( Very intersting .. Interesting that it start crashes now with LIBLG_RECYCLEFBO only after fix, so like when all things start to works correclty it then behave badly..

**Comment by kas1e** [comment]
I do search in google for "irrlicht FBO has one or several incomplete image attachements", and found some links from which i understand that seem irrlicht before quering on the possible attributes, and then use those ones which is found, like floating point textures, and all other stuff.

Maybe, it just found on our driver something which make it behave like that (and which, maybe didn't find on pandora? ) 

Another man wrote this "I discovered that if you create a render target with a color format which is not supported you'll have this FBO error, however nothing more (no crash). BUT, the next render target you will create will crash the app."

I then do some more google, and found, that one of user in 2012 can't run supertuxkart with the same errors , but it was on Ubuntu, and ppls answered him that he probabaly have issues with drivers.

I also found the same mention of such error on MineTest's forum, and the answers also was  like "its your hardware or drivers or you don't have enough video memory" , etc , etc.

**Comment by ptitSeb** [comment]
Nah... Remember you had a "trace" of the framebuffer that where created (with the `DEBUG` define), and the FBO created where regular one, with a simple textue (ok, BGRA, but even forcing to RGBA doesn't change anything).

**Comment by kas1e** [comment]
I also can think about me messing around of pathces, but then, if old, non-fixed LIBGL_RECYCLEFBO didn't crash, then it mean all fine.

I just for now catch again this crash, but this time i can see rendered car in process of crash. It can mean nothing, or can mean something, but see there is back wheel not rendered fully. But that can of course can be because it just didn't have time to do so because crashes:

http://kas1e.mikendezign.com/aos4/gl4es/games/supertuxkart_081/fbo_crash.jpg

**Comment by ptitSeb** [comment]
Before my fix on LIBGL_RECYCLEFBO, a recycled FBO was not always binded correctly. So if no FBO were binded, there is no FBO, no wonder it worked ok.

**Comment by kas1e** [comment]
Understood.. I for now made an archive for Daniel with all the tracelogs, so now only to wait when he find the issue.

In meantime i meassuer loading speed on different computers, and:

1). my amiga (ssd hdd, sata) : 100 seconds
2). icorei5 2.7 ghz (1 year old notebook): 12 seconds

So i then i start to check wtf, run our dos tracer, and found that i have a looooot of native GetFilePosition()/ChangeFilePosition() calls on b3d files, when that slow loading happens (it start to be slow once 10st pict is shown on loading screen, after "red" head, when "karts/konqi" loading starts). 

Then i check Irrlicht/CReadFile.cpp, and it use fseek()/ftell() and stuff, so i test them to see, when 
GetFilePosition()/ChangeFilePosition() called , and its ftell() , which by some reassons are very slow on our side when it come to some .b3d files. 

Now i recheck with dos-tracer, what happens when that "karts/konqi" loading starts, and it indeed all those calls happens for long:

```
05637 : supertuxkart    : o.k. = GetFilePosition(0x18CF484C [Work:supertuxkart/data/karts/konqi/konqi.b3d]) = 1125556 [2uS]
05638 : supertuxkart    : o.k. = GetFilePosition(0x18CF484C [Work:supertuxkart/data/karts/konqi/konqi.b3d]) = 1158324 [3uS]
05639 : supertuxkart    : o.k. = ChangeFilePosition(0x18CF484C [Work:supertuxkart/data/karts/konqi/konqi.b3d],1125564,OFFSET_BEGINNING) [2uS]
05640 : supertuxkart    : o.k. = GetFilePosition(0x18CF484C [Work:supertuxkart/data/karts/konqi/konqi.b3d]) = 1125564 [2uS]
05641 : supertuxkart    : o.k. = GetFilePosition(0x18CF484C [Work:supertuxkart/data/karts/konqi/konqi.b3d]) = 1158332 [3uS]
```

And that take a minute.  I checked that file, and its almost 20 mb of size , and seems ftell() there (our implementation at least , seems so), suck hard. 

Then i found another part where slowdown on loading happens, it is "karts/sara", so with the same lots of calls (i assume its ftell() ).  And that "karts/sara" also have b3d file about 8mb of size. There slowing down exactly on 50% less as with "konqi" 19mb file , so it kind of tell us that problem is big (well, not BIG, but bigger than others) loading of .b3d files  when ftell() is used on them.

For sake of tests i just remove from data/karts those 2 directories : "konqi" and "sara" , and loading take just 15 seconds.

Very intersting .. Now need to understand what happens , and why there lot of ftells(), and what they expected to do .. Maybe it possible to just replace them out, to not use newlib's ftell() somehow. If that of course, ftell() issue.

**Comment by ptitSeb** [comment]
It is pretty strange that a "ftell" is so demanding. Is there some reason for that? Can you use something else then `FILE` that has faster random access maybe?

**Comment by kas1e** [comment]
I can't say with 100% sure that its ftell, but all i can see that when loading of such a  "big" b3d files happens, i have lots of native functions which probabaly mean ftell(). Maybe there some "buffering" happens or something which need to be disabled ..

> Can you use something else then FILE that has faster random access maybe?

Dunno, i read somehwere that can be some memory mapping used, like you load file to memory , and then. But probably i firstly need to create some tests case which will take that "big" b3d file, and  
doing the same what irrlicht do (but what it do ? why there needs for looots of ftells() ? )

**Comment by ptitSeb** [comment]
Creating a memory buffer is quite easy indeed. But changing all file access in b3d reader for Irrlicht is maybe more difficult.

**Comment by ptitSeb** [comment]
On another subject: I have seen a french article of Stunt Car Remake on AmigaOS4 & MorphOS here: http://obligement.free.fr/articles/stuntcarremake.php

I know your are not the author of the port, but I was wondering if you can get in touch with him. It seems the port has issues. I see in the test that the engine tend to disapear, and that the "large screen" cockpit I drawn is also not shown. I don't have any of this issue.
I know the game is not finished yet (only train mode, where the damage reset and you cannot choose your oponent), but those technical are not in current version of the game. I you have a Linux or Windows browser, you can even try it inline here: http://ptitseb.github.io/stuntcarremake/

**Comment by kas1e** [comment]
Sure i will try, but i didn't know him , and if i remember right, there was some "legal" issues or something and games disappear from being availlable or something like that (if i don't mess it with another racing game of late). He for sure not vocal visitor of amiga sites, so i even didn't know at moment who he is :) But i will try to check/ask/etc of course.

As for ftell() : i thinking at first maybe it possible to replace ftell() only, like call ftell_my() instead of newlib's ftell.. 

What i probabaly need, its somehow measure ftell() speed  : on amigaos4 we have 2 c libs, newlib (which i use) and clib2 (which i offten avoid, but which can be better there). So will be interesting to make some test case which i can measure with different c libs.

**Comment by ptitSeb** [comment]
Ok, thanks (for StuntCarRemake).

For your slowyness in B3D Reading, you can try this diff:
```
diff --git a/source/Irrlicht/CB3DMeshFileLoader.cpp b/source/Irrlicht/CB3DMeshFileLoader.cpp
old mode 100644
new mode 100755
index 3d8769e1..7ac8a199
--- a/source/Irrlicht/CB3DMeshFileLoader.cpp
+++ b/source/Irrlicht/CB3DMeshFileLoader.cpp
@@ -14,6 +14,10 @@
 #include "IVideoDriver.h"
 #include "IFileSystem.h"
 #include "os.h"
+#define MEMORY
+#ifdef MEMORY
+#include "CMemoryFile.h"
+#endif

 #ifdef _DEBUG
 #define _B3D_READER_DEBUG
@@ -51,8 +55,15 @@ IAnimatedMesh* CB3DMeshFileLoader::createMesh(io::IReadFile* f)
 {
        if (!f)
                return 0;
-
+#ifndef MEMORY
        B3DFile = f;
+#else
+       long sz = f->getSize();
+       c8* buff = new c8[sz];
+       f->read(buff, sz);
+       B3DFile = new io::CMemoryReadFile(buff, sz, f->getFileName(), true);
+       f->seek(0);
+#endif
        AnimatedMesh = new scene::CSkinnedMesh();
        ShowWarning = true; // If true a warning is issued if too many textures are used
        VerticesStart=0;
@@ -66,6 +77,10 @@ IAnimatedMesh* CB3DMeshFileLoader::createMesh(io::IReadFile* f)
                AnimatedMesh->drop();
                AnimatedMesh = 0;
        }
+#ifdef MEMORY
+       delete B3DFile;
+       B3DFile = nullptr;
+#endif

        return AnimatedMesh;
 }
```
Not sure it works, but it should buffer the whole b3d in memory before reading. If the slow comes from ftell / fseek, it should speed up stuff...

**Comment by kas1e** [comment]
Oh, thanks a bunch, that will be good test now for sure ! 

Through it failed to compile because of that line:
```B3DFile = new io::CMemoryReadFile(buff, sz, f->getFileName(), true);```

says that ``io``` are guilty with words error: expected type-specifier

**Comment by kas1e** [comment]
Ah its probabaly CMemoryReadFile : found in changelog.txt:

 - (Internal) Replaced CMemoryReadFile with CMemoryFile, that also implements an IWriteFile interface.

**Comment by kas1e** [comment]
Ok, with your path and just changed CMemoryReadFile on CMemoryFile, it now didn't take even few seconds ! 

I mean, i have now 2 exactly the same versions : one original irrlicht's code, and another one with your path. That one which is original, take 100 seconds on loading. With your patch - 15 seconds maximum. Like, all slownes somewhere disappear ! Veeerry strange !

Speed of loading now the same as on my win10 on icore5. Even maybe a little bit faster. But as i can see, that because on win32 , ftell() and co still slow things down even on win32 -> theere is noticable stop for a few seconds when loading "korni" and "sara". 

Can you also try it on Pandora ? Will it the same radically speed up things for you ? I can't test for a moment exactly "korni" and "sara" cars (as i offten crash on car-choice screen), but so far it seems just a MEGA difference, for Pandora it will be for sure cool as well.

**Comment by ptitSeb** [comment]
> Ah its probabaly CMemoryReadFile : found in changelog.txt:
> 
> * (Internal) Replaced CMemoryReadFile with CMemoryFile, that also implements an IWriteFile interface.

Ah, I use a slightly older version of Irrlicht it seems...



> Can you also try it on Pandora ? Will it the same radically speed up things for you ? I can't test for a moment exactly "korni" and "sara" cars (as i offten crash on car-choice screen), but so far it seems just a MEGA difference, for Pandora it will be for sure cool as well.

Well, I'll try later (need some time to relink, because of lto). Also, Memory Buffer are of course much faster than file, espacialy on random access. But that much seems suspicious: you still might have some functions that are much slower than what they should be on your AmigaOS4.

**Comment by kas1e** [comment]
Very possible that all FILE functions of newlib are that bad. Or maybe just ftell(). I just create simple test case like this:

```

#include<stdio.h> 
  
int main() 
{ 
    /* Opening file in read mode */
    FILE *fp = fopen("test.txt","r"); 
  
    /* Reading first string */
    char string[20];    
    fscanf(fp,"%s",string); 
  
    /* Printing position of file pointer */
    printf("%ld", ftell(fp)); 
    return 0; 
}
```

Then create test.txt with 2 bytes inside, and the compile it in 2 versions: newlib and clib2. And for newlib dos-tracer output are:

```
00027 : ftell_newlib    : o.k. = Open("test.txt",OLD) = [0x1984643C] [22uS]
00028 : ftell_newlib    : o.k. = ExamineObject(0x6316E848 [test.txt]) [16uS]
00029 : ftell_newlib    : FAIL = IsInteractive("test.txt")
00030 : ftell_newlib    : o.k. = IsFileSystem("") [4uS]
00031 : ftell_newlib    : o.k. = ExamineObject(0x6316E7A8 [test.txt]) [12uS]
00032 : ftell_newlib    : o.k. = GetFilePosition(0x1984643C [test.txt]) = 3 [3uS]
00033 : ftell_newlib    : o.k. = GetFilePosition(0x1984643C [test.txt]) = 3 [2uS]
00034 : ftell_newlib    : o.k. = ChangeFilePosition(0x1984643C [test.txt],2,OFFSET_BEGINNING) [3uS]
00035 : ftell_newlib    : o.k. = GetFilePosition(0x1984643C [test.txt]) = 2 [1uS]
00036 : ftell_newlib    : FAIL = ExamineObject(0x6316E538 [<null>]) [4617uS]
00037 : ftell_newlib    :        SetCurrentDir("<untracked>")
00038 : ftell_newlib    : DIR  = ParentOfFH(0x1984643C) "test.txt" [5uS]
00039 : ftell_newlib    : o.k. = ExamineObject(0x6316EC78 [test.txt]) [12uS]
00040 : ftell_newlib    : <----- RunCommand(0x18E89915 "ftell_newlib",,"",1) = 0 [0x00000000] [21304uS]
```

And for clib2 are:

```
00027 : ftell_clib2     : o.k. = Open("test.txt",OLD) = [0x19ECFD14] [18uS]
00028 : ftell_clib2     : FAIL = IsInteractive("test.txt")
00029 : ftell_clib2     : o.k. = IsFileSystem("") [3uS]
00030 : ftell_clib2     : DIR  = ParentOfFH(0x19ECFD14) "test.txt" [5uS]
00031 : ftell_clib2     : o.k. = ExamineFH("test.txt") [459uS]
00032 : ftell_clib2     : <----- RunCommand(0x18E89915 "ftell_clib2",,"",1) = 0 [0x00000000] [4949uS]
```

And while both are working correctly, and return me "2" in console, from log can be seen that newlib's version doing 4 native calls which clib2 didn't do ! I.e. those: 

GetFileposition()
GetFileposition()
ChangeFileposition()
GetFileposition()

Bunch of which i see when i run supertuxkart with original FILE usage of ftell() , and which actually slow things down.

I think, that just newlib's ftell() slow like crap and that the only real explain.

Because for example loading of those 20mb files to memory , etc all happens fast (with all fopen() and co). Its just ftell() seems reacts like this. I do not remember if any other game/app use ftell() quite heavy , maybe its just irrlicht/supertuxkart are first ones i meet with such issue.

**Comment by ptitSeb** [comment]
What are the uS number in square brakets? Time in µ seconds or something else. In that case, the call itself is fast (2 or 3 µs for get position call seems ok, no?).

About SuperTuxKart, you now you can disable FBO in the config (or probably in config menu too) to try see how the game play on the amiga.

**Comment by kas1e** [comment]
> What are the uS number in square brakets? Time in µ seconds or something else. In that case, the call itself is fast (2 or 3 µs for get position call seems ok, no?).

Probabaly microseconds , yes, but maybe issues is not speed of call itself, but amount of those calls ?

> About SuperTuxKart, you now you can disable FBO in the config (or probably in config menu too) to try see how the game play on the amiga.

Rigth, i didn't crash anymore when choice a car, but its slooowert :) Not THAT slower, but visibly. Through no more crash :)

Inside of levels all fine as before by speed.

But i stil have some hardcore strange crashes when trying to load some maps , but that probabaly we can keep for later, once Daniel with deal with FBO crashes (as maybe even when i disable them, its somehow related).

**Comment by ptitSeb** [comment]
Mmmm odd there are still crash. They should not be FBO related.

Ok, lets wait FBO fix then.

**Comment by ptitSeb** [comment]
I tried SuperTuxKart with the hacked code for B3D loader: it load fast on the Pandora too. But even without the hack, it was still loading quite fast anyway, so not much changes here.
(but the last gl4es improvements, with VBO and optimized shaders, have a sensitive impact on STK on the Pandora, most race are smooth enough now)

**Comment by kas1e** [comment]
I enabled all the tracks (in challenge.html, change all false on true). And those ones which crash on me all the time (some hardcore luckup):

```
  - BlackHill Mansion (data/tracks/mansion/)
  - Old Main (data/tracks/mines/)
  - STK Enterprise (data/tracks/starttrack/)
  - Zen Garden (data/tracks/zengarden/)
```

Can you check plz, if all of those works for you on Pandora ?

**Comment by ptitSeb** [comment]
> So that just prove that ftell() on our side is slow very much and that your change indeed works fine.
> 
> As for whole STK gameplay, are you able let's say, to play few tracks, then exit, and start "tutorial mode" without crashes , etc ?

No it doesn't. It can only proves that some or all of file management functions (`fseek`, `ftell` and `fread` at least) are much slower than they should be. But it's not necessarily `ftell`



> I enabled all the tracks (in challenge.html, change all false on true). And those ones which crash on me all the time (some hardcore luckup):
> 
> ```
>   - BlackHill Mansion (data/tracks/mansion/)
>   - Old Main (data/tracks/mines/)
>   - STK Enterprise (data/tracks/starttrack/)
>   - Zen Garden (data/tracks/zengarden/)
> ```
> 
> Can you check plz, if all of those works for you on Pandora ?

I'm not home anymore. I'll try tonight. Before leaving I tried 2 tracks (school-thmed and the one underwater) then go to story mode, and it work. I did ran out of VRAM on the Pandora, but that's not unusual for STK (and I'm not sure how much VRAM I have, probably around 32MB, but there is no way, to my knowledge, to tell how much total VRAM or how much free VRAM there is in GLES2).

**Comment by kas1e** [comment]
I just retested all the tracks, and yep, it crashes only on those 4 ones. And also "story mode" crashes too, most of time at least. 

I found, that if i remove some cars directories from data/karts, then i can have "story mode" working. Sometime i even can start "BlackHill Mansion". That probabaly point out on some memory filling/trashing somewhere .. Just crash is so heavy, that it didn't bring any kind of stack trace , everything just locked up, and i didn't even have any idea where to look :(  All what i was able to capture, its when i use debug version of RadeonHD driver and Warp3DNova.library is that:

```
RadeonHD.chip (5): rhdRMCtxUnlockBufferInternal called (0x57E4D318)
RadeonHD.chip (5): rhdRMCtxUnlockBufferInternal done.
RadeonHD.chip (5): rhdRMCtxCopyToVRAM called
RadeonHD.chip (5): rhdRMCtxLockBuffer called (0x57E4D318)
RadeonHD.chip (5): rhdRMCtxUnlockBuffer called (0x57E4D318, 1104319)
RadeonHD.chip (5): rhdRMCtxCopyToVRAM done
W3DN_SI.library (6): </ctxTexUpdateSubImage result="0x00000000">
W3DN_SI.library (6): </ctxTexUpdateImage result="0x00000000">
W3DN_SI.library (6): <ctxBindTexture>
W3DN_SI.library (6): </ctxBindTexture result="0x00000000">
W3DN_SI.library (6): <ctxTexUpdateImage>
W3DN_SI.library (6): <ctxTexUpdateSubImage>
RadeonHD.chip (5): rhdRMCtxUnlockBufferInternal called (0x57E4D318)
RadeonHD.chip (5): rhdRMCtxUnlockBufferInternal done.
RadeonHD.chip (5): rhdRMCtxCopyToVRAM called
RadeonHD.chip (5): rhdRMCtxLockBuffer called (0x57E4D318)
RadeonHD.chip (5): rhdRMCtxUnlockBuffer called (0x57E4D318, 1104320)
RadeonHD.chip (5): rhdRMCtxCopyToVRAM done
W3DN_SI.library (6): </ctxTexUpdateSubImage result="0x00000000">
W3DN_SI.library (6): </ctxTexUpdateImage result="0x00000000">
W3DN_SI.library (6): <ctxBindTexture>
W3DN_SI.library (6): </ctxBindTexture result="0x00000000">
W3DN_SI.library (6): <ctxCreateTexSampler>
W3DN_SI.library (6): </ctxCreateTexSampler result="0x55682B68">
W3DN_SI.library (6): <ctxTSSetParameters>
W3DN_SI.library (6): </ctxTSSetParameters result="0x00000000">
W3DN_SI.library (6): <ctxTSSetParameters>
W3DN_SI.library (6): </ctxTSSetParameters result="0x00000000">
RadeonHD.chip (5): rhdRMCtxUnlockBufferInternal called (0x57E4D318)
RadeonHD.chip (5): rhdRMCtxUnlockBufferInternal done.
W3DN_SI.library (6): <ctxCreateTexture>
RadeonHD.chip (5): rhdRMCtxAllocBuffer called
RadeonHD.chip (5): rhdRMReadExecutionID called
RadeonHD.chip (4): executionID: 1104323
RadeonHD.chip (5): rhdRMMAllocIntBuffer called
W3DN_SI.library (6): </ctxCreateTexture result="0x5567EAC8">
W3DN_SI.library (6): <ctxTexSetParameters>
W3DN_SI.library (6): </ctxTexSetParameters result="0x00000000">
W3DN_SI.library (6): <ctxBindTexture>
W3DN_SI.library (6): </ctxBindTexture result="0x00000000">
W3DN_SI.library (6): <ctxTexSetParameters>
W3DN_SI.library (6): </ctxTexSetParameters result="0x00000000">
W3DN_SI.library (6): <ctxTexUpdateImage>
W3DN_SI.library (6): <ctxTexUpdateSubImage>
RadeonHD.chip (5): rhdRMCtxCopyToVRAM called
RadeonHD.chip (5): rhdRMCtxLockBuffer called (0x57E4D2B8)
RadeonHD.chip (5): rhdRMCtxUnlockBuffer


[HAL_DfltTrapHandler] *** Warning: Fatal exception in task 0x65FC93A0 (Shell Process, etask = 0xEFD3D6C0) at ip 0x01E51048
```
So all the time those strange lockups happens, it happens when radeonhd driver tried to "rhdRMCtxUnlockBuffer" , after warp3dnova do one of many ctxTexUpdateSubImage.

I also tried to catch just ogles2 trace:
http://kas1e.mikendezign.com/aos4/gl4es/games/supertuxkart_081/lockup/ogles2_trace.txt

And just warp3dnova trace:
http://kas1e.mikendezign.com/aos4/gl4es/games/supertuxkart_081/lockup/nova_trace.txt

Trying to understand wtf happens

**Comment by ptitSeb** [comment]
How much VRAM do you have? Maybe it's Out of VRAM issue. Maybe try with `LIBGL_SHRINK=2` to shrink textures (and reduce vram used), or even with `LIBGL_USEVBO=0` to again reduce vram used. Just for trying.

**Comment by kas1e** [comment]
I have 2gb of video memory , when i crashes i have 1535mb free. But i tried anyway those options, and while just LIBGL_USEVBO 0 change nothing, LIBGL_SHRINK 2 change A LOT ! Now, i almost never crashes. I can load all those levels, but after a while i still crashes.

Then i tried LIBGL_SHRINK 4 , and that one NEVER crashes does not matter what i do and in what tracks i play. 

Then i start to check by sysmonitor, what happens, and seems that when used GPU memory is more than 249mb (so probably when it hit 256mb barier), then i have those issues. 

Need experement more, but at moment it looks very much like RadeonHD driver issue then.

**Comment by kas1e** [comment]
So for me currently "uncrashable" state are LIBGL_SHRINK 4, where i never hit 250mb of used GPU.

One time i can see that 258mb was used before lockup. I.e. anything close to "256mb" of used GPU for sure lockup.

**Comment by kas1e** [comment]
Oh and LIBGL_SHRINK 1 reduce memory usage very much, never hit even 100mb of used graphics memory with that, and things still continue to looks more or less good

**Comment by ptitSeb** [comment]
Ok VRAM issue then.

Look in USAGE.md for detail about all the shrink mode. Mode 1 use a very crude shrinking algorithm, but yeah, it will reduce memory used by texture by 4. On Pandora, I used mode 2 IIRC (but without VBO, so maybe now I need mode 4?).

**Comment by kas1e** [comment]
I can say that 0.8.1 version of supertuxkart works the same by speed as with and same without VBO (at least, that what i notice). And even if you disable VBO, it still didn't reduce usage of video memory.

But i need to recheck that

**Comment by ptitSeb** [comment]
Speed is the same? Mmmm, strange, I do have a slight speedup with VBO. Also, no change in memory consuption is also strange. VBO should be in VRAM. But maybe it's in a different part and/or not counted in your memory counter?

**Comment by kas1e** [comment]
Oh sorry i was wrong. Tested just some level , and when VBO is 1 , it give me ~36 fps, but when VBO is 0, it give me ~26fps.  

But in terms of video memory, without VBO, i have on some level 210mb of video memory used, but WITH vbo, i have 212mb of video memory used. 

So at least in terms of saving video memory usage usage of VBO didn't make a lot of differences. Maybe because, that even without VBO in gl4es, we sitll use internally our client->vram vbos in ogles2, which almost on pair as real vbo in client (but still slower)

**Comment by ptitSeb** [comment]
Ah ok, and yes, there will be VBO anyway, so the memory wont change much (and VBO must be smaller then texture anyway).

**Comment by kas1e** [comment]
Maybe in you will be in interest about, but i did more tests about our VRAM issue, and found that i can crash it easyl just with such a test case:

```
#include <stdlib.h>
#include <stdio.h>
#include <proto/exec.h>
#include <proto/intuition.h>
#include <GLES2/gl2.h>
#include <proto/ogles2.h>


struct Library *OGLES2 = NULL;
struct OGLES2IFace *IOGLES2 = NULL;

void texturesload_crash(void)
{


// 100 : eat 34 mb of GPU 
// 200 : eat 68 mb of GPU 
// 300 : eat 101 mb of GPU 
// 400 : eat 135 mb of GPU 
// 500 : eat 168 mb of GPU 
// 600 : eat 202 mb of GPU 
// 700 : eat 235 mb of GPU 
// 750 : eat 252 mb of GPU 

// 764 : fill 256 mb of GPU , but not overbound to 257 mb at moment 
// 765 : now we cross the line of 256 mb => CRASH 


    int num_of_textures = 764;

    int width = 256;
    int height = 256;
    unsigned char *data;
    data = malloc(width*height*3);    

    GLuint TextureID[num_of_textures];
        
    for(int a=0;a<num_of_textures;a++)
    {
        glGenTextures( 1, &TextureID[a]);
        glBindTexture( GL_TEXTURE_2D, TextureID[a]);
        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, width, height,0, GL_RGB, GL_UNSIGNED_BYTE, data);
        
    }

    free(data);


}



/*  Main Loop
 *  Open window with initial window size, title bar,
 *  RGBA display mode, and handle input events.
 */
int main(int argc, char** argv)
{

        OGLES2 = IExec->OpenLibrary("ogles2.library",0);
        IOGLES2 = (struct OGLES2IFace *)IExec->GetInterface(OGLES2, "main", 1, NULL);
        


        ULONG errCode = 0;
           const char titleStrBase[] = "Hello OpenGLES2";
        struct Window *win=IIntuition->OpenWindowTags(NULL,
                                WA_Title,                titleStrBase,
                                WA_Activate,            TRUE,
                                WA_RMBTrap,                TRUE,
                                WA_DragBar,                TRUE,
                                WA_DepthGadget,            TRUE,
                                WA_SimpleRefresh,        TRUE,
                                WA_CloseGadget,            TRUE,
                                WA_SizeGadget,            TRUE,
                                WA_IDCMP,                IDCMP_REFRESHWINDOW | IDCMP_NEWSIZE | 
                                                        IDCMP_CLOSEWINDOW | IDCMP_RAWKEY,
                                WA_InnerWidth,            500,
                                WA_InnerHeight,            200,
                                WA_MinWidth,            100,
                                WA_MinHeight,            100,
                                WA_MaxWidth,            2048,
                                WA_MaxHeight,            2048,
                                WA_BackFill,             LAYERS_NOBACKFILL,
                                TAG_DONE);
        
    struct TagItem contextparams[] =
    {
            {OGLES2_CCT_WINDOW,(ULONG)win},
            {OGLES2_CCT_DEPTH,16},
            {OGLES2_CCT_STENCIL,8},
            {OGLES2_CCT_VSYNC,0},
            {OGLES2_CCT_SINGLE_GET_ERROR_MODE,1},
            {OGLES2_CCT_RESIZE_VIEWPORT, TRUE},
            {TAG_DONE,0}
    };
        
    void *ogles_context = aglCreateContext2(&errCode,contextparams);

      if (ogles_context)
    {
        aglMakeCurrent(ogles_context);

        texturesload_crash();
        
        BOOL bRun = TRUE;

        if (!IIntuition->ModifyIDCMP(win, IDCMP_VANILLAKEY|IDCMP_MOUSEMOVE|IDCMP_MOUSEBUTTONS|IDCMP_NEWSIZE|IDCMP_CLOSEWINDOW))
            bRun = FALSE;

        while (bRun)
        {
            struct IntuiMessage *imsg;


            aglSwapBuffers();

            while ((imsg = (struct IntuiMessage *)IExec->GetMsg(win->UserPort)))
            {
                switch (imsg->Class)
                {
                    case IDCMP_CLOSEWINDOW:
                        bRun = FALSE;
                        break;
                    case IDCMP_VANILLAKEY:
                        switch (imsg->Code)
                        {
                            case 27:
                                aglDestroyContext(ogles_context);
                                IIntuition->CloseWindow(win);
                                exit(0);
                                break;
                        }
                        
                    default:
                        break;
                }
                IExec->ReplyMsg((struct Message *)imsg);
                }
        }
        aglDestroyContext(ogles_context);
        IIntuition->CloseWindow(win);
    }    
    return 0;
}
```

So no GL4ES involved, no SDL ,  just pure ogles2 + warp3dnova. 

As can be seen in comments, once i fill first 256 mb of memory and didn't cross the 256mb barier then all fine, system operational , etc. But once i cross the 256mb barier, and want just a little bit from 257's mb then BAH !  

So that "256" mb crash looks very "programming bug" issue, and while you of course can't know how our driver done, maybe you know is there some "technique" how developers of drirvers works with GPU usually ? Maybe its the same good-old bank switching used and VRAM splitted per 256mb blocks ?

**Comment by ptitSeb** [comment]
Well, I can't help you much there. It's all in Hans's hand I suppose. I guess there an issue or a limit in the GPU memory to CPU memory mapping...

**Comment by kas1e** [comment]
While it for sure not gl4es at all realted, but maybe you still in interst to know about progress about that bug : i do another test today which cost me 200$ : i buy RadeonRX 570 8 GB (Polaris10) , and we have driver for that too (RadeonRX.chip). So, with that one, i can fill up 4gb of GPU video memory EASY without crashes/lockups.  That mean, that its probabaly not hardware limit of GPU<->CPU memory mapping, as if it works on the same hardware but with different driver, then it's probabaly RadeonHD.chip driver have some bug.

**Comment by ptitSeb** [comment]
Ah, new fancy graphic card? Nice, I wonder if some game are faster with that one or not (if not, that would means the GPU is alway waiting for the CPU to complete tasks).

Yeah, the bug is limited to RadeonHD driver (can still be about GPU<->CPU mapping).

**Comment by kas1e** [comment]
As i aware, on other amigaone hardware (not x5000 which i have, but amiga x1000 & SAM460) , there is no such a bug with RadeonHD, but they use some "Dma cpu hack" in graphics.library, which very well maybe related to GPU<->CPU mapping and just "workaround" that issue. 

As for new card, it currently only works for me for 2D and for very basic 3d (like to fill gpu data, some little demo), once anything heaver, driver then bugs. So i need to wait for proper use it fixed driver first (probabaly card is too new for our driver, as i just go to shop and buy fresh one).

**Comment by ptitSeb** [comment]
Ah I see. And all this is only handled by Hans?! That's a lot for a single man. No wonder he is overworked!

**Comment by kas1e** [comment]
Yeah, he do RadeonRH, RadeonRX drivers (which mean 2d, compisiting, dvi/hdmi outputs , etc), Warp3DNova (so whole 3D driver, with shader's compiler, etc) , then he also lately made some pleminar support of hardware video decoding in RadeonRX, then he soon hope to works on GART,  as well as lately he made some ftp server with TLS support , in other words, he now completely burted out, and take a vacation for month probabaly. But as i aware he on solary from Trevor (that one you meet on gamecom party), so its kind of "work" for him.

**Comment by kas1e** [comment]
Btw, just to let you know, that issue with FILE operation (i.e. that slow loading of supertuxkart when its fread/ftell/etc used), was because of old Newlib's code, where ftell() have fflush() inside, which just kill all the buffered reading.

Now our coder who works on Newlib, backport newer code from latest 3.1.0 newlib where is no fflush() and all that slow stuff, and everything works fine and fast now, almost the same as from RAM:

**Comment by ptitSeb** [comment]
Ah, good to hear. That should have helped some other games I guess, like Barony for example.

**Comment by kas1e** [comment]
Btw, wasnt' you tried to port more fresh version of SuperTuxKart ? I know they now use OpenGL3.X (so we out of luck there probabaly), but maybe some previous version than 1.0, but after 0.8.1 (when then change irrlicht on their own engine) can works too ? Never tried ?

**Comment by ptitSeb** [comment]
No I haven't tried 0.9.x
I know they also have native GLES2 and GLES3 support, but the texture and model are even more heavy, so I didn't bother on the Pandora.

You can try, maybe the GLES2 version can work fine on Amiga.

**Comment by kas1e** [comment]
Interesting ! 

Firstly let they fix it all for 0.8.1 so all will works, and then can check newer ones. But good that it have GLES2 support, pretty neat !

**Comment by kas1e** [comment]
Btw, Hans fix some shader issue which prevent before supertuxkart shaders to be compiled, and not when i set LIBGL_NPOT to 2 , its all starts , but i can't find where that "post processing blur effect" are visibly in game, are you remember where ?

**Comment by kas1e** [comment]
Oh sorry, i read you previous comment where you say it should be when use turbo, but then .. i really didn't see a difference, be it with shaders or without, probabaly the same look :)

**Comment by ptitSeb** [comment]
The shader trigger when using Turbo (you have to callect some before using, or do some slide to gain a bit of turbo boost), or when drive just behind an opponent and get some dragging effect from him.

**Comment by kas1e** [comment]
Well even if it ther, i really didn't see much difference, so not big worry for sure then.

**Comment by kas1e** [comment]
Probably we can close that one in the meantime as well as at least for amigaos4 everything is done. Check this out how is it: https://www.youtube.com/watch?v=uyuT4UGfDc8

**Comment by ptitSeb** [comment]
Ah yeah, it's working pretty well! Looks really enjoyable on the Amiga like this :)

**Comment by kas1e** [comment]
Just to inform: with the latest GL4ES FBO crashes are gone when choosing a car with enabled FBO in the game :) And loading time because of our recent fix about textures reduced from 17 seconds to just 6 :)

**Comment by ptitSeb** [comment]
Sounds good :)
